// @azure/cognitiveservices-anomalydetector/dist/cognitiveservices-anomalydetector.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@azure/ms-rest-js')) :
    typeof define === 'function' && define.amd ? define(['exports', '@azure/ms-rest-js'], factory) :
    (global = global || self, factory((global.Azure = global.Azure || {}, global.Azure.CognitiveservicesAnomalydetector = {}), global.msRest));
}(this, (function (exports, msRest) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /*
     * Copyright (c) Microsoft Corporation. All rights reserved.
     * Licensed under the MIT License. See License.txt in the project root for license information.
     *
     * Code generated by Microsoft (R) AutoRest Code Generator.
     * Changes may cause incorrect behavior and will be lost if the code is regenerated.
     */

    var index = /*#__PURE__*/Object.freeze({
        __proto__: null
    });

    /*
     * Copyright (c) Microsoft Corporation. All rights reserved.
     * Licensed under the MIT License. See License.txt in the project root for license information.
     *
     * Code generated by Microsoft (R) AutoRest Code Generator.
     * Changes may cause incorrect behavior and will be lost if the code is regenerated.
     */
    var APIError = {
        serializedName: "APIError",
        type: {
            name: "Composite",
            className: "APIError",
            modelProperties: {
                code: {
                    serializedName: "code",
                    type: {
                        name: "Object"
                    }
                },
                message: {
                    serializedName: "message",
                    type: {
                        name: "String"
                    }
                }
            }
        }
    };
    var Point = {
        serializedName: "Point",
        type: {
            name: "Composite",
            className: "Point",
            modelProperties: {
                timestamp: {
                    required: true,
                    serializedName: "timestamp",
                    type: {
                        name: "DateTime"
                    }
                },
                value: {
                    required: true,
                    serializedName: "value",
                    type: {
                        name: "Number"
                    }
                }
            }
        }
    };
    var Request = {
        serializedName: "Request",
        type: {
            name: "Composite",
            className: "Request",
            modelProperties: {
                series: {
                    required: true,
                    serializedName: "series",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Composite",
                                className: "Point"
                            }
                        }
                    }
                },
                granularity: {
                    required: true,
                    nullable: false,
                    serializedName: "granularity",
                    type: {
                        name: "Enum",
                        allowedValues: [
                            "yearly",
                            "monthly",
                            "weekly",
                            "daily",
                            "hourly",
                            "minutely",
                            "secondly"
                        ]
                    }
                },
                customInterval: {
                    serializedName: "customInterval",
                    type: {
                        name: "Number"
                    }
                },
                period: {
                    serializedName: "period",
                    type: {
                        name: "Number"
                    }
                },
                maxAnomalyRatio: {
                    serializedName: "maxAnomalyRatio",
                    type: {
                        name: "Number"
                    }
                },
                sensitivity: {
                    serializedName: "sensitivity",
                    type: {
                        name: "Number"
                    }
                }
            }
        }
    };
    var EntireDetectResponse = {
        serializedName: "EntireDetectResponse",
        type: {
            name: "Composite",
            className: "EntireDetectResponse",
            modelProperties: {
                period: {
                    required: true,
                    serializedName: "period",
                    type: {
                        name: "Number"
                    }
                },
                expectedValues: {
                    required: true,
                    serializedName: "expectedValues",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Number"
                            }
                        }
                    }
                },
                upperMargins: {
                    required: true,
                    serializedName: "upperMargins",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Number"
                            }
                        }
                    }
                },
                lowerMargins: {
                    required: true,
                    serializedName: "lowerMargins",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Number"
                            }
                        }
                    }
                },
                isAnomaly: {
                    required: true,
                    serializedName: "isAnomaly",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Boolean"
                            }
                        }
                    }
                },
                isNegativeAnomaly: {
                    required: true,
                    serializedName: "isNegativeAnomaly",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Boolean"
                            }
                        }
                    }
                },
                isPositiveAnomaly: {
                    required: true,
                    serializedName: "isPositiveAnomaly",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Boolean"
                            }
                        }
                    }
                }
            }
        }
    };
    var LastDetectResponse = {
        serializedName: "LastDetectResponse",
        type: {
            name: "Composite",
            className: "LastDetectResponse",
            modelProperties: {
                period: {
                    required: true,
                    serializedName: "period",
                    type: {
                        name: "Number"
                    }
                },
                suggestedWindow: {
                    required: true,
                    serializedName: "suggestedWindow",
                    type: {
                        name: "Number"
                    }
                },
                expectedValue: {
                    required: true,
                    serializedName: "expectedValue",
                    type: {
                        name: "Number"
                    }
                },
                upperMargin: {
                    required: true,
                    serializedName: "upperMargin",
                    type: {
                        name: "Number"
                    }
                },
                lowerMargin: {
                    required: true,
                    serializedName: "lowerMargin",
                    type: {
                        name: "Number"
                    }
                },
                isAnomaly: {
                    required: true,
                    serializedName: "isAnomaly",
                    type: {
                        name: "Boolean"
                    }
                },
                isNegativeAnomaly: {
                    required: true,
                    serializedName: "isNegativeAnomaly",
                    type: {
                        name: "Boolean"
                    }
                },
                isPositiveAnomaly: {
                    required: true,
                    serializedName: "isPositiveAnomaly",
                    type: {
                        name: "Boolean"
                    }
                }
            }
        }
    };
    var ChangePointDetectRequest = {
        serializedName: "ChangePointDetectRequest",
        type: {
            name: "Composite",
            className: "ChangePointDetectRequest",
            modelProperties: {
                series: {
                    required: true,
                    serializedName: "series",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Composite",
                                className: "Point"
                            }
                        }
                    }
                },
                granularity: {
                    required: true,
                    nullable: false,
                    serializedName: "granularity",
                    type: {
                        name: "Enum",
                        allowedValues: [
                            "yearly",
                            "monthly",
                            "weekly",
                            "daily",
                            "hourly",
                            "minutely",
                            "secondly"
                        ]
                    }
                },
                customInterval: {
                    serializedName: "customInterval",
                    type: {
                        name: "Number"
                    }
                },
                period: {
                    serializedName: "period",
                    type: {
                        name: "Number"
                    }
                },
                stableTrendWindow: {
                    serializedName: "stableTrendWindow",
                    type: {
                        name: "Number"
                    }
                },
                threshold: {
                    serializedName: "threshold",
                    type: {
                        name: "Number"
                    }
                }
            }
        }
    };
    var ChangePointDetectResponse = {
        serializedName: "ChangePointDetectResponse",
        type: {
            name: "Composite",
            className: "ChangePointDetectResponse",
            modelProperties: {
                period: {
                    required: true,
                    serializedName: "period",
                    type: {
                        name: "Number"
                    }
                },
                isChangePoint: {
                    required: true,
                    serializedName: "isChangePoint",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Boolean"
                            }
                        }
                    }
                },
                confidenceScores: {
                    required: true,
                    serializedName: "confidenceScores",
                    type: {
                        name: "Sequence",
                        element: {
                            type: {
                                name: "Number"
                            }
                        }
                    }
                }
            }
        }
    };

    var Mappers = /*#__PURE__*/Object.freeze({
        __proto__: null,
        APIError: APIError,
        Point: Point,
        Request: Request,
        EntireDetectResponse: EntireDetectResponse,
        LastDetectResponse: LastDetectResponse,
        ChangePointDetectRequest: ChangePointDetectRequest,
        ChangePointDetectResponse: ChangePointDetectResponse
    });

    /*
     * Copyright (c) Microsoft Corporation. All rights reserved.
     * Licensed under the MIT License. See License.txt in the project root for
     * license information.
     *
     * Code generated by Microsoft (R) AutoRest Code Generator.
     * Changes may cause incorrect behavior and will be lost if the code is
     * regenerated.
     */
    var endpoint = {
        parameterPath: "endpoint",
        mapper: {
            required: true,
            serializedName: "Endpoint",
            defaultValue: '',
            type: {
                name: "String"
            }
        },
        skipEncoding: true
    };

    /*
     * Copyright (c) Microsoft Corporation. All rights reserved.
     * Licensed under the MIT License. See License.txt in the project root for
     * license information.
     *
     * Code generated by Microsoft (R) AutoRest Code Generator.
     * Changes may cause incorrect behavior and will be lost if the code is
     * regenerated.
     */
    var packageName = "@azure/cognitiveservices-anomalydetector";
    var packageVersion = "2.1.0";
    var AnomalyDetectorClientContext = /** @class */ (function (_super) {
        __extends(AnomalyDetectorClientContext, _super);
        /**
         * Initializes a new instance of the AnomalyDetectorClientContext class.
         * @param endpoint Supported Cognitive Services endpoints (protocol and hostname, for example:
         * https://westus2.api.cognitive.microsoft.com).
         * @param credentials Subscription credentials which uniquely identify client subscription.
         * @param [options] The parameter options
         */
        function AnomalyDetectorClientContext(credentials, endpoint, options) {
            var _this = this;
            if (endpoint == undefined) {
                throw new Error("'endpoint' cannot be null.");
            }
            if (credentials == undefined) {
                throw new Error("'credentials' cannot be null.");
            }
            if (!options) {
                options = {};
            }
            if (!options.userAgent) {
                var defaultUserAgent = msRest.getDefaultUserAgentValue();
                options.userAgent = packageName + "/" + packageVersion + " " + defaultUserAgent;
            }
            _this = _super.call(this, credentials, options) || this;
            _this.baseUri = "{Endpoint}/anomalydetector/v1.0";
            _this.requestContentType = "application/json; charset=utf-8";
            _this.endpoint = endpoint;
            _this.credentials = credentials;
            return _this;
        }
        return AnomalyDetectorClientContext;
    }(msRest.ServiceClient));

    /*
     * Copyright (c) Microsoft Corporation. All rights reserved.
     * Licensed under the MIT License. See License.txt in the project root for
     * license information.
     *
     * Code generated by Microsoft (R) AutoRest Code Generator.
     * Changes may cause incorrect behavior and will be lost if the code is
     * regenerated.
     */
    var AnomalyDetectorClient = /** @class */ (function (_super) {
        __extends(AnomalyDetectorClient, _super);
        /**
         * Initializes a new instance of the AnomalyDetectorClient class.
         * @param endpoint Supported Cognitive Services endpoints (protocol and hostname, for example:
         * https://westus2.api.cognitive.microsoft.com).
         * @param credentials Subscription credentials which uniquely identify client subscription.
         * @param [options] The parameter options
         */
        function AnomalyDetectorClient(credentials, endpoint, options) {
            return _super.call(this, credentials, endpoint, options) || this;
        }
        AnomalyDetectorClient.prototype.entireDetect = function (body, options, callback) {
            return this.sendOperationRequest({
                body: body,
                options: options
            }, entireDetectOperationSpec, callback);
        };
        AnomalyDetectorClient.prototype.lastDetect = function (body, options, callback) {
            return this.sendOperationRequest({
                body: body,
                options: options
            }, lastDetectOperationSpec, callback);
        };
        AnomalyDetectorClient.prototype.changePointDetect = function (body, options, callback) {
            return this.sendOperationRequest({
                body: body,
                options: options
            }, changePointDetectOperationSpec, callback);
        };
        return AnomalyDetectorClient;
    }(AnomalyDetectorClientContext));
    // Operation Specifications
    var serializer = new msRest.Serializer(Mappers);
    var entireDetectOperationSpec = {
        httpMethod: "POST",
        path: "timeseries/entire/detect",
        urlParameters: [
            endpoint
        ],
        requestBody: {
            parameterPath: "body",
            mapper: __assign(__assign({}, Request), { required: true })
        },
        responses: {
            200: {
                bodyMapper: EntireDetectResponse
            },
            default: {
                bodyMapper: APIError
            }
        },
        serializer: serializer
    };
    var lastDetectOperationSpec = {
        httpMethod: "POST",
        path: "timeseries/last/detect",
        urlParameters: [
            endpoint
        ],
        requestBody: {
            parameterPath: "body",
            mapper: __assign(__assign({}, Request), { required: true })
        },
        responses: {
            200: {
                bodyMapper: LastDetectResponse
            },
            default: {
                bodyMapper: APIError
            }
        },
        serializer: serializer
    };
    var changePointDetectOperationSpec = {
        httpMethod: "POST",
        path: "timeseries/changePoint/detect",
        urlParameters: [
            endpoint
        ],
        requestBody: {
            parameterPath: "body",
            mapper: __assign(__assign({}, ChangePointDetectRequest), { required: true })
        },
        responses: {
            200: {
                bodyMapper: ChangePointDetectResponse
            },
            default: {
                bodyMapper: APIError
            }
        },
        serializer: serializer
    };

    exports.AnomalyDetectorClient = AnomalyDetectorClient;
    exports.AnomalyDetectorClientContext = AnomalyDetectorClientContext;
    exports.AnomalyDetectorMappers = Mappers;
    exports.AnomalyDetectorModels = index;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
